gce: Prefer MASTER_ADVERTISE_ADDRESS in apiserver setup

grosskur opened the pull request:
MASTER_ADVERTISE_ADDRESS is used to set the --advertise-address flag
for the apiserver. It's useful for running the apiserver behind a load
balancer.

However, if PROJECT_ID, TOKEN_URL, TOKEN_BODY, and NODE_NETWORK are
all set, the GCE VM's external IP address will be fetched and used
instead and MASTER_ADVERTISE_ADDRESS will be ignored.

Change this behavior so that MASTER_ADVERTISE_ADDRESS takes precedence
because it's more specific. We still fall back to using the VM's
external IP address if the other variables are set.

Also: Move the setting of --ssh-user and --ssh-keyfile based on
PROXY_SSH_USER) to a top-level block because this is common to all
codepaths.

```release-note
GCE: Fix to make the built-in `kubernetes` service properly point to the master's load balancer address in clusters that use multiple master VMs.
```

grosskur committed:
gce: Prefer MASTER_ADVERTISE_ADDRESS in apiserver setup

MASTER_ADVERTISE_ADDRESS is used to set the --advertise-address flag
for the apiserver. It's useful for running the apiserver behind a load
balancer.

However, if PROJECT_ID, TOKEN_URL, TOKEN_BODY, and NODE_NETWORK are
all set, the GCE VM's external IP address will be fetched and used
instead and MASTER_ADVERTISE_ADDRESS will be ignored.

Change this behavior so that MASTER_ADVERTISE_ADDRESS takes precedence
because it's more specific. We still fall back to using the VM's
external IP address if the other variables are set.

Also: Pass --ssh-user and --ssh-keyfile flags if both PROXY_SSH_USER
and MASTER_ADVERTISE_ADDRESS is set.

grosskur commented:
I've now signed the CNCF CLA.

bowei commented:
/ok-to-test

bowei commented:
/assign

bowei commented:
/assign @dnardo

dnardo commented:
/lgtm other than Bowei's comments.

Was kubeclt logs/exec tested with this change (in all cases) ?

grosskur commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -1622,15 +1622,15 @@ function start-kube-apiserver {
+    params+=" --advertise-address=${MASTER_ADVERTISE_ADDRESS}"
+  elif [[ -n "${PROJECT_ID:-}" && -n "${TOKEN_URL:-}" && -n "${TOKEN_BODY:-}" && -n "${NODE_NETWORK:-}" ]]; then
+    local -r vm_external_ip=$(curl --retry 5 --retry-delay 3 ${CURL_RETRY_CONNREFUSED} --fail --silent -H 'Metadata-Flavor: Google' "http://metadata/computeMetadata/v1/instance/network-interfaces/0/access-configs/0/external-ip")
+    params+=" --advertise-address=${vm_external_ip}"
```
Not sure. I looked at the git history but couldn't figure out why --advertise-address was only set if PROXY_SSH_USER was set. To be safe, I've updated the diff to leave that entire block as-is.

grosskur commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -1622,15 +1622,15 @@ function start-kube-apiserver {
-    fi
-  elif [ -n "${MASTER_ADVERTISE_ADDRESS:-}" ]; then
-    params="${params} --advertise-address=${MASTER_ADVERTISE_ADDRESS}"
+  if [ -n "${MASTER_ADVERTISE_ADDRESS:-}" ]; then
```
Done.

grosskur commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -1622,15 +1622,15 @@ function start-kube-apiserver {
+  if [ -n "${MASTER_ADVERTISE_ADDRESS:-}" ]; then
+    params+=" --advertise-address=${MASTER_ADVERTISE_ADDRESS}"
+  elif [[ -n "${PROJECT_ID:-}" && -n "${TOKEN_URL:-}" && -n "${TOKEN_BODY:-}" && -n "${NODE_NETWORK:-}" ]]; then
+    local -r vm_external_ip=$(curl --retry 5 --retry-delay 3 ${CURL_RETRY_CONNREFUSED} --fail --silent -H 'Metadata-Flavor: Google' "http://metadata/computeMetadata/v1/instance/network-interfaces/0/access-configs/0/external-ip")
```
Whoops. I initially developed this patch on an older branch that didn't have `get-metadata-value`. Switched to use `get-metadata-value` now.

grosskur commented:
Thanks for the review!

@bowei: I've pushed a new diff that addresses your comments.

@dnardo: I've tested both `kubectl exec` and `kubectl logs` in all cases now, and verified they still work.

grosskur commented:
/retest

grosskur commented:
/test pull-kubernetes-integration

bowei commented:
/approve no-issue

bowei commented:
/lgtm
